### ISSUE TITLE
Using API requires impl.Processor outside API package

### ISSUE THREAD

**Issue by Reoseah** [swordglowsblue/artifice#1]
https://github.com/swordglowsblue/artifice/blob/1c63ea1ba8b42c3eb48aac809e66757bb2a09924/src/main/java/com/swordglowsblue/artifice/api/ArtificeResourcePack.java#L8

Imo classes that user of library will require should all be in `api` package. But currently one has to pass `Processor` to `ArtificeResourcePack#addBlockState` and similar methods in that class (although most of the time that would be lambda function and it becomes not obvious that something is in `impl` package). 

Also it isn't clear for me why it needs to return something in the first place, all methods in ArtificeResourcePack could use standard `Consumer`, removing need to constantly write `return builder;`.

**Comment by swordglowsblue** [comment]
`Processor` is designed as a "returning consumer", if you will; it's a passthrough similar to Kotlin's `.apply {}` that prevents dozens or hundreds of lines like the following:

```java
Builder builder = new Builder();
consumer.accept(builder);
return builder.build();
```

Instead, the following can be used:

```java
return processor.process(new Builder()).build();
```

`Processor` is ubiquitous throughout even Artifice's internal code, so you're probably right that it should be exposed in the `api` package. I originally placed it in `impl` since it should generally be invisible to the user when working with lambdas, as I've been very careful to ensure that all builders are fully chaineable and can be written in a single expression (removing the need to return explicitly from a lambda). I'm not really very inclined to refactor hundreds of lines of code to replace it with `Consumer`, though.

**Comment by Reoseah** [comment]
Well, it works well when using this style of lambdas
```java
Artifice.registerAssets("operatingpower:virtual", ArtificeResourcePack.ofAssets(pack -> {
			String[] blocks = { "copper_block" };
			for (String block : blocks) {
				pack.addBlockState(id(block), state -> state
						.variant("", settings -> settings
								.model(id(block))));
				pack.addBlockModel(id(block), model -> model
						.parent(new Identifier("minecraft:block/cube_all"))
						.texture("all", id("block/" + block)));
				pack.addItemModel(id(block), model -> model
						.parent(id("block/" + block)));
			}
		}));
```
but with `{}` returns must be at the end of each lambda...
```java
Artifice.registerAssets("operatingpower:virtual", ArtificeResourcePack.ofAssets(pack -> {
			String[] blocks = { "copper_block" };
			for (String block : blocks) {
				pack.addBlockState(id(block), state -> {
					state.variant("", settings -> {
						settings.model(id(block));
						return settings;
					});
					return state;
				});
				pack.addBlockModel(id(block), model -> {
					model.parent(new Identifier("minecraft:block/cube_all"))
							.texture("all", id("block/" + block));
					return model;
				});
				pack.addItemModel(id(block), model -> {
					model.parent(id("block/" + block));
					return model;
				});
			}
		}));
``` 
In theory, api can be overloaded to have both `Processor` and `Consumer` variants...

**Comment by swordglowsblue** [comment]
I think I may add a `Processor.of(Consumer)` static method that allows for easy wrapping. Overloading the entire API is a lot of work, but it would theoretically be manageable with that, or at worst it could be a slightly less redundant-looking bit of boilerplate.

**Comment by Reoseah** [comment]
not sure but this may work
```java
@FunctionalInterface
public interface Processor<T> extends Consumer<T> {
    default T process(T t) {
    	return t;
    }
}
```

**Comment by swordglowsblue** [comment]
Thanks, that works beautifully =) Closed by f5c767b, will be in 0.3.0.
